### PR TITLE
Allow using CustomTarget as test executable

### LINKED ISSUE #6567: How to use an executable made by a custom_target() as command for test() ?
with meson 0.53.1
I've a custom_target() that generate an executable.
I want to give it as the command to execute for the test() command.
How can I do that (if possible)?

### PR DESCRIPTION
**Pull request by bruchar1**
Fixes #6567

### CODE REVIEW

File: mesonbuild/interpreter/interpreter.py [hunk 1]
```diff
@@ -2153,17 +2153,17 @@ def func_generator(self, node: mparser.FunctionNode,
         self.generators.append(gen)
         return gen
 
-    @typed_pos_args('benchmark', str, (build.Executable, build.Jar, ExternalProgram, mesonlib.File))
+    @typed_pos_args('benchmark', str, (build.Executable, build.Jar, ExternalProgram, mesonlib.File, build.CustomTarget, build.CustomTargetIndex))
```
Comment (dcbaker) [inline mesonbuild/interpreter/interpreter.py:2156]: We should have a `FeatureNew` here

File: mesonbuild/interpreter/interpreter.py [hunk 2]
```diff
@@ -2153,17 +2153,17 @@ def func_generator(self, node: mparser.FunctionNode,
...
     @typed_kwargs('benchmark', *TEST_KWS)
     def func_benchmark(self, node: mparser.BaseNode,
                        args: T.Tuple[str, T.Union[build.Executable, build.Jar, ExternalProgram, mesonlib.File]],
                        kwargs: 'kwtypes.FuncBenchmark') -> None:
         self.add_test(node, args, kwargs, False)
 
-    @typed_pos_args('test', str, (build.Executable, build.Jar, ExternalProgram, mesonlib.File))
+    @typed_pos_args('test', str, (build.Executable, build.Jar, ExternalProgram, mesonlib.File, build.CustomTarget, build.CustomTargetIndex))
```
Comment (dcbaker) [inline mesonbuild/interpreter/interpreter.py:2163]: and here

**Review by dcbaker — CHANGES_REQUESTED** [review]
Apart from missing `FeatureNew` checks, this looks good to me

File: docs/markdown/snippets/test_using_custom_target_executable.md [hunk 3]
```diff
@@ -0,0 +1,4 @@
+## Use `custom_target` as test executable
+
+[[test]] function now accepts [[@custom_tg]] and [[@custom_idx]] for the
```
Comment (eli-schwartz) [inline docs/markdown/snippets/test_using_custom_target_executable.md:3]: Missing "The" at the beginning of the sentence.

Second link should fail the "update website" CI.

File: mesonbuild/mtest.py [hunk 4]
```diff
@@ -1456,6 +1456,11 @@ def _get_test_cmd(self) -> T.Optional[T.List[str]]:
                            'found. Please check the command and/or add it to PATH.')
                     raise TestException(msg.format(self.test.exe_wrapper.name))
                 return self.test.exe_wrapper.get_command() + self.test.fname
+        elif self.test.cmd_is_built and is_windows():
+            test_cmd = ExternalProgram._shebang_to_cmd(self.test.fname[0])
+            if test_cmd is not None:
+                test_cmd += self.test.fname[1:]
+            return test_cmd
```
Comment (eli-schwartz) [inline mesonbuild/mtest.py:1463]: How do we handle File and ExternalProgram here? This feels like we are not doing something earlier that we should have done...

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: mesonbuild/mtest.py (hunk 4 above)
Comment (eli-schwartz) [reply mesonbuild/mtest.py:1463]: Ugh, makes sense.

**Review by dcbaker — APPROVED** [review]

**Pull request merged**